pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java

```diff
@@ -989,6 +1016,12 @@ protected void doTakeSnapshot() {
       }
     }
     _updatedSegmentsSinceLastSnapshot.clear();
+    // Persist TTL watermark after taking snapshots if TTL is enabled, so that segments out of TTL can be loaded with
+    // updated validDocIds bitmaps. If the TTL watermark is persisted first, segments out of TTL may get loaded with
+    // stale bitmaps or even no bitmap snapshots to use.
+    if (isTTLEnabled()) {
+      persistWatermark(_largestSeenComparisonValue.get());
+    }
```

Comment:
Same this will be persisting watermark for only deletedKeysTTL enabled table which is not necessary

Reply:
for my understanding, why watermark is not needed for only deletedKeysTTL enabled table?

Reply:
There is no `outOfTTLSegment` concept in deletedKeysTTL flow. We add all the segments during server-start phase and calculate `_largestComparisonValue` while adding these segments. We only remove keys which are **marked for delete** and out of deletedKeysTTL threshold.

Reply:
as discussed offline, there is no harm to persist/restore watermark for table just using deletedKeysTTL, and considering when deletedKeysTTL and metadataTTL are both enabled for a table, deletedKeysTTL uses the restored watermark anyway, so I kept this logic to make tracking watermark consistent for both TTL types.